downloads/downloads-impl/src/main/java/com/duckduckgo/downloads/impl/DefaultFileDownloadNotificationManager.kt

```diff
@@ -21,69 +21,112 @@ import android.app.PendingIntent
+        waitForNotificationToBeCancelled(downloadId)
+        notificationManager.notify(downloadId.toInt(), notification)
     }
 
-    override fun showDownloadFailedNotification() {
-        mainThreadHandler().post {
-            val notification = NotificationCompat.Builder(applicationContext, FileDownloadNotificationChannelType.FILE_DOWNLOADED.id)
-                .setContentTitle(applicationContext.getString(R.string.downloadFailed))
-                .setSmallIcon(R.drawable.ic_file_download_white_24dp)
-                .build()
+    @AnyThread
+    override fun showDownloadFailedNotification(downloadId: Long, url: String?) {
+        val notification = NotificationCompat.Builder(applicationContext, FileDownloadNotificationChannelType.FILE_DOWNLOADED.id)
+            .setContentTitle(applicationContext.getString(R.string.downloadFailed))
+            .setSmallIcon(R.drawable.ic_file_download_white_24dp)
+            .apply {
+                url?.let { fileUrl ->
+                    val pendingIntent = PendingIntent.getBroadcast(
+                        applicationContext,
+                        1234,
+                        FileDownloadNotificationActionReceiver.retryDownloadIntent(downloadId, fileUrl),
+                        PendingIntent.FLAG_UPDATE_CURRENT
+                    )
+                    addAction(R.drawable.ic_file_download_white_24dp, applicationContext.getString(R.string.downloadsRetry), pendingIntent)
+                }
+            }
+            .build()
 
-            notificationManager.notify(NOTIFICATION_ID, notification)
+        // cancelling notifications is a deferred action, and I have seen how in subsequent notify calls end up in the wrong end state
+        notificationManager.cancel(downloadId.toInt())
+        waitForNotificationToBeCancelled(downloadId)
+        notificationManager.notify(downloadId.toInt(), notification)
+    }
+
+    @AnyThread
+    override fun cancelDownloadFileNotification(downloadId: Long) {
+        notificationManager.cancel(downloadId.toInt())
+    }
+
+    private fun createIntentToOpenFile(applicationContext: Context, file: File): Intent {
+        val fileUri = getFilePathUri(applicationContext, file)
+        return Intent().apply {
+            setDataAndType(fileUri, applicationContext.contentResolver?.getType(fileUri))
+            action = Intent.ACTION_VIEW
+            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_GRANT_READ_URI_PERMISSION
         }
     }
 
-    private fun mainThreadHandler() = Handler(Looper.getMainLooper())
+    private fun getFilePathUri(context: Context, file: File): Uri {
+        return FileProvider.getUriForFile(context, "${appBuildConfig.applicationId}.provider", file)
+    }
 
-    companion object {
-        private const val NOTIFICATION_ID = 1
+    private fun waitForNotificationToBeCancelled(downloadId: Long) {
```

Comment:
You can remove this if you use unique request codes as most probably this is why you see the issue mentioned in the comment.

Reply:
Not sure I understand this comment. Let's discuss it offline

Reply:
Removed this in one of the latest commits